### PR TITLE
Increase performance for conversions including axis angles

### PR DESCRIPTION
**Pull request by alex-bene**
This is an extension of #1544 with various speed, stability, and readability improvements. (I could not find a way to make a commit to the existing PR). This PR is still based on the [Rodrigues' rotation formula](https://en.wikipedia.org/wiki/Rotation_formalisms_in_three_dimensions#Rotation_matrix_%E2%86%94_Euler_axis/angle).

The motivation is the same; this change speeds up the conversions up to 10x, depending on the device, batch size, etc.

### Notes
- As the angles get very close to `π`, the existing implementation and the proposed one start to differ. However, (my understanding is that) this is not a problem as the axis can not be stably inferred from the rotation matrix in this case in general.
- @bottler , I tried to follow similar conventions as existing functions to deal with weird angles, let me know if something needs to be changed to merge this.

### CODE REVIEW

**Comment by alex-bene** [comment]
@bottler let's continue the conversation here.
Regarding your comment on #1544 , do you want to add the "fast" argument to both matrix->axis angle and axis angle-> matrix conversions? Also, all the rest of the conversions (e.g. `quaternion_to_axis_angle`) use a fixed `eps=1e-6`. While I do agree this should be an input argument, this is different from the existing convention. Is it okay to add this argument to all functions that use it?

**Comment by bottler** [comment]
Yes let's add the "fast" to both cases. And I don't mind the eps being inconsistent: actually feel free to add it everywhere or in these two places or nowhere.

**Comment by bottler** [comment]
Yes let's add the "fast" to both cases. And I don't mind the eps being inconsistent: actually feel free to add it everywhere or in these two places or nowhere.

**Comment by bottler** [comment]
Yes please! I think the change to torch.sinc would be a good idea. It wasn't tried before. Thanks!

**Comment by bottler** [comment]
Prefer torch.special.sinc to torch.sinc .

**Comment by alex-bene** [comment]
`torch.special.sinc` is the same as `torch.sinc` as far as I understand.
Also, FYI, `sinc` generates the exact same results as a naive implementation of `sinx/x`:

```python
import torch

naive = lambda x: torch.sin(x) / x if x != 0 else 1

res = torch.tensor([
    naive(i) - torch.sinc(i/torch.pi).item()
    for i in torch.linspace(0, 1e-6, 100000)
])

assert res.abs().max() == 0.0
assert res.abs().mean() == 0.0
```

**Comment by bottler** [comment]
So the whole taylor series thing probably wasn't useful at all!

**Comment by alex-bene** [comment]
> So the whole taylor series thing probably wasn't useful at all!

It seems so

So, I spent quite a lot of time to figure out what to do in the conversion `matrix_to_axis_angle` when the angle is near `pi` or `-pi`. Eventually, it seems like the solution is to do an intermediate conversion to quaternions.

In total, I feel quite confident to set `fast=True` in general, but take a look and let me know if there's anything else that seems like it could create a problem.

**Comment by bottler** [comment]
> So, I spent quite a lot of time to figure out what to do in the conversion `matrix_to_axis_angle` when the angle is near `pi` or `-pi`. Eventually, it seems like the solution is to do an intermediate conversion to quaternions.

In which case, is this change to that function actually a speed improvement?

**Comment by alex-bene** [comment]
> In which case, is this change to that function actually a speed improvement

Well, considering a uniform distribution of angles, this should still be a speed improvement (unless all angles were near pi for example).

However, calling the functions for the intermediate conversion even with empty tables led to very slow performance (not sure why). So, I just committed a change to avoid this. Now there is a total improvement in speed across all cases **plus** the function is stable in all edge cases.

**Comment by alex-bene** [comment]
Hey @bottler , I can see that some tests are falling, however, I do not have access to see exactly what fails. Is it possible to share the error logs to fix this?

**Comment by bottler** [comment]
Thanks for this work! Tests are fine. Probably you saw some reflection of longstanding failures.

**Comment by alex-bene** [comment]
Okay, I see. Glad I could help. 🙌
